### DISCUSSION TITLE
My Game Contract

### DISCUSSION THREAD

**Discussion by JoshuyPan** [Cyfrin/foundry-full-course-cu#2098] [Ideas]
Hey Guys!

Actually i'm on Section 4: Lottery Contract, btw i was scratching on Remix and i shaped this "Game", with this structs:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

contract Classes {
    
    enum ObjectRarity{
        Trash,
        Common,
        UnCommon,
        Rare,
        Legendary,
        Artifact
    }

    struct Object{
        bytes32 objName;
        uint256 objId;
        ObjectRarity rarity;
        uint256 value;
        Chib owner;
    }

    struct Chib{
        address ownerAddr;
        uint256 level;
        uint256 hp;
        uint256 mana;
        Object[] inventory;
    }

    struct User {
        string userName;
        Chib[] characters;
        bool isOnline;
        address payable addr;
    }    

}
```

and here my new user function:

```solidity
    function registerNewUser(string memory userName) external onlyOneRegistration {
        string memory check = userName.toLowerCase();
        for (uint256 i = 0; i < s_users.length; i++) {
            string memory checkExist = s_users[i].userName.toLowerCase();
            if (keccak256(bytes(check)) == keccak256(bytes(checkExist))) {
                revert Game__UsernameAlreadyExists();
            }
        }
        //If so, create the new user, push it into users array and into mapping
        Chib[] memory emptyCharactersArray;
        User memory user = User(
            userName,
            emptyCharactersArray,
            true,
            payable(msg.sender)
        );
        s_users.push(user);
        addressToUser[msg.sender] = user;
        emit NewUser(msg.sender, user.userName);
    }
```
How can i actually initialize the Chibs[] when i cast the user for the first time below?
        );
```solidity
        Chib[] memory emptyCharactersArray;
        User memory user = User(
            userName,
            emptyCharactersArray, //<--- HERE
            true,
            payable(msg.sender)
        );
```

because as i'm doing, i receive this error:

```
UnimplementedFeatureError: Copying of type struct Classes.Chib memory[] memory to storage not yet supported.
```

I know i'm doing something wrong but i can't reconize what is it. Actually the AI Buddy suggested me to do it in this way, but it doesn't work.

**Comment by EngrPips** [comment]
This seems interesting, and I will advise you to consult the solidity docs  `https://docs.soliditylang.org/en/v0.8.26/`

**Comment by Cryptoineer** [comment]
I think you need to restructure you code or change the architecture to make it work. Here are some example to make it work:

```
// SPDX-License-Identifier: MIT

pragma solidity 0.8.24;

contract Storage {

    enum ObjectRarity {
        Trash,
        Common,
        UnCommon,
        Rare,
        Legendary,
        Artifact
    }

    struct Object {
        bytes32 objName;
        uint256 objId;
        ObjectRarity rarity;
        uint256 value;
        address owner; // Store owner address directly in Object
    }

    struct Chib {
        address ownerAddr;
        uint256 level;
        uint256 hp;
        uint256 mana;
    }

    struct User {
        string userName;
        bool isOnline;
        address payable addr;
    }

    // Mappings
    mapping(address => User) public users;
    mapping(address => Chib[]) public userCharacters; // Map user address to an array of Chibs
    mapping(address => mapping(uint256 => Object[])) public characterInventory; // Map user address, character index to an array of Objects

    function registerNewUser(string memory userName) external {
        User memory newUser = User({
            userName: userName,
            isOnline: true,
            addr: payable(msg.sender)
        });
        users[msg.sender] = newUser;

        // Create initial character (you can customize this logic)
        Chib memory newChib = Chib({
            ownerAddr: msg.sender,
            level: 1,
            hp: 100,
            mana: 50
        });
        userCharacters[msg.sender].push(newChib);
    }

    function addObjectToCharacterInventory(address userAddress, uint256 characterIndex, Object memory newObject) public {
        newObject.owner = userAddress; // Set owner address in the Object
        characterInventory[userAddress][characterIndex].push(newObject);
    }

    // Other functions to interact with characters, objects, etc.
}
```

**Reply by EngrPips** [reply]
Interesting

**Comment by JoshuyPan** [comment]
> I think you need to restructure you code or change the architecture to make it work. Here are some example to make it work:
> 
> 
> 
> ```
> 
> // SPDX-License-Identifier: MIT
> 
> 
> 
> pragma solidity 0.8.24;
> 
> 
> 
> contract Storage {
> 
> 
> 
>     enum ObjectRarity {
> 
>         Trash,
> 
>         Common,
> 
>         UnCommon,
> 
>         Rare,
> 
>         Legendary,
> 
>         Artifact
> 
>     }
> 
> 
> 
>     struct Object {
> 
>         bytes32 objName;
> 
>         uint256 objId;
> 
>         ObjectRarity rarity;
> 
>         uint256 value;
> 
>         address owner; // Store owner address directly in Object
> 
>     }
> 
> 
> 
>     struct Chib {
> 
>         address ownerAddr;
> 
>         uint256 level;
> 
>         uint256 hp;
> 
>         uint256 mana;
> 
>     }
> 
> 
> 
>     struct User {
> 
>         string userName;
> 
>         bool isOnline;
> 
>         address payable addr;
> 
>     }
> 
> 
> 
>     // Mappings
> 
>     mapping(address => User) public users;
> 
>     mapping(address => Chib[]) public userCharacters; // Map user address to an array of Chibs
> 
>     mapping(address => mapping(uint256 => Object[])) public characterInventory; // Map user address, character index to an array of Objects
> 
> 
> 
>     function registerNewUser(string memory userName) external {
> 
>         User memory newUser = User({
> 
>             userName: userName,
> 
>             isOnline: true,
> 
>             addr: payable(msg.sender)
> 
>         });
> 
>         users[msg.sender] = newUser;
> 
> 
> 
>         // Create initial character (you can customize this logic)
> 
>         Chib memory newChib = Chib({
> 
>             ownerAddr: msg.sender,
> 
>             level: 1,
> 
>             hp: 100,
> 
>             mana: 50
> 
>         });
> 
>         userCharacters[msg.sender].push(newChib);
> 
>     }
> 
> 
> 
>     function addObjectToCharacterInventory(address userAddress, uint256 characterIndex, Object memory newObject) public {
> 
>         newObject.owner = userAddress; // Set owner address in the Object
> 
>         characterInventory[userAddress][characterIndex].push(newObject);
> 
>     }
> 
> 
> 
>     // Other functions to interact with characters, objects, etc.
> 
> }
> 
> ```

Thanks for your answer!
So there is no chance to store an array of structs inside a struct?

**Comment by Cryptoineer** [comment]
Refer this link to get more clarity : https://www.reddit.com/r/solidity/comments/xlgwk1/how_to_add_an_array_of_structs_as_a_parameter_in/

**Reply by JoshuyPan** [reply]
> Refer this link to get more clarity : https://www.reddit.com/r/solidity/comments/xlgwk1/how_to_add_an_array_of_structs_as_a_parameter_in/
> 
> 

Thanks 🙏🏻
